### PR TITLE
[SPARK-42372][SQL] Improve performance of HiveGenericUDTF by making inputProjection instantiate once

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Improve performance of HiveGenericUDTF by making inputProjection instantiate once

### Why are the changes needed?

performance improvement for hive udtf

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

via benchmarking, over 2x faster with one LOC.

```gitdiff
+++ b/sql/hive/benchmarks/HiveUDFBenchmark-per-row-results.txt
@@ -0,0 +1,7 @@
+OpenJDK 64-Bit Server VM 1.8.0_352-bre_2022_12_13_23_06-b00 on Mac OS X 13.1
+Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
+Hive UDTF benchmark:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------------------------------
+Hive UDTF dup 2                                    1574           1680         118          0.7        1501.1       1.0X
+Hive UDTF dup 4                                    2642           3076         588          0.4        2519.9       0.6X
+
diff --git a/sql/hive/benchmarks/HiveUDFBenchmark-results.txt b/sql/hive/benchmarks/HiveUDFBenchmark-results.txt
new file mode 100644
index 0000000000..8af8b6582c
--- /dev/null
+++ b/sql/hive/benchmarks/HiveUDFBenchmark-results.txt
@@ -0,0 +1,7 @@
+OpenJDK 64-Bit Server VM 1.8.0_352-bre_2022_12_13_23_06-b00 on Mac OS X 13.1
+Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
+Hive UDTF benchmark:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
+------------------------------------------------------------------------------------------------------------------------
+Hive UDTF dup 2                                     712            789         101          1.5         678.7       1.0X
+Hive UDTF dup 4                                    1212           1294          78          0.9        1156.0       0.6X
+ 
```

### CODE REVIEW

**Review by wangyum — APPROVED** [review]

File: sql/hive/benchmarks/HiveUDFBenchmark-results.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+OpenJDK 64-Bit Server VM 1.8.0_352-bre_2022_12_13_23_06-b00 on Mac OS X 13.1
```
Comment (LuciferYang) [inline sql/hive/benchmarks/HiveUDFBenchmark-results.txt:1]: @yaooqinn please add Java 11 & 17 result too.

Comment (LuciferYang) [reply sql/hive/benchmarks/HiveUDFBenchmark-results.txt:1]: And the benchmark result should be produced by GA

Comment (wangyum) [reply sql/hive/benchmarks/HiveUDFBenchmark-results.txt:1]: May be we can remove the benchmark?

Comment (yaooqinn) [reply sql/hive/benchmarks/HiveUDFBenchmark-results.txt]: +1 to @wangyum 's suggestion, as after this fix, the benchmarking is mostly evaluating hive udf execution

Comment (LuciferYang) [reply sql/hive/benchmarks/HiveUDFBenchmark-results.txt]: fine to remove the benchmark

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Comment by yaooqinn** [comment]
thanks, merged to master/3.4

**Pull request closed without merging**